### PR TITLE
Sipa cancel membership

### PR DESCRIPTION
**Pull request by ibot3**
Makes Pycroft ready for terminating memberships via SIPA.

### CODE REVIEW

File: pycroft/lib/finance.py [hunk 1]
```diff
@@ -776,16 +776,77 @@ def match_activities():
...
+
+
+def estimate_balance(user, end_date):
+    """
+    :param user: The member
+    :param end_date: Date of the end of the membership
+    :return: Estimated balance at the end_date
+    """
+
+    now = session.utcnow().date() + timedelta(1)
```
Comment (Georgeto) [inline pycroft/lib/finance.py]: What's the reasoning for counting from the next day instead of from today?

File: tests/lib/test_finance.py [hunk 2]
```diff
@@ -304,3 +313,95 @@ def test_ordered(self):
...
+        self.user = UserFactory.create()
+
+        self.user_membership = MembershipFactory.create(
+            begins_at=session.utcnow() - timedelta(weeks=52),
+            ends_at=None,
+            user=self.user,
+            group=config.member_group
+        )
+
+        last_month_last = session.utcnow().date().replace(day=1) - timedelta(1)
```
Comment (Georgeto) [inline tests/lib/test_finance.py]: Maybe add a function `last_day_of_prev_month` analogous to `last_day_of_month`?

File: pycroft/lib/finance.py [hunk 3]
```diff
@@ -776,16 +776,77 @@ def match_activities():
...
+    end_date_justified = end_date - timedelta(last_fee.booking_begin.days - 1)
+
+    months_to_pay = diff_month(end_date_justified, now)
+
+    # If the user has to pay a fee for the current month
+    if user.has_property('membership_fee',
+                         single(now.replace(day=last_fee.booking_end.days))):
+        months_to_pay += 1
+
+    last_month_last = now.replace(day=1) - timedelta(1)
```
Comment (Georgeto) [inline pycroft/lib/finance.py]: Maybe add a function `last_day_of_prev_month` analogous to `last_day_of_month`?

File: pycroft/lib/finance.py [hunk 4]
```diff
@@ -776,16 +776,77 @@ def match_activities():
...
+    current_month_last = last_day_of_month(now)
+
+    # Get the split for a maybe existent membership fee booking in the current
+    # month
+    split_current_month = (Split.q
+                           .filter_by(account=user.account)
+                           .join(Transaction)
+                           .filter(Split.amount > 0)
+                           .filter(Transaction.valid_on == current_month_last)
+                           ).first()
```
Comment (lukasjuhrich) [inline pycroft/lib/finance.py]: Might be extracted, is precisely the same as `split_last_month` modulovariation by `current_month_last`.

File: pycroft/helpers/date.py [hunk 5]
```diff
@@ -0,0 +1,10 @@
+from datetime import timedelta
+
+
+def diff_month(d1, d2):
+    return (d1.year - d2.year) * 12 + d1.month - d2.month
```
[the remark this answers is not in the document]
Comment (lukasjuhrich) [reply pycroft/helpers/date.py]: Now there are. B-)

File: web/api/v0/__init__.py [hunk 6]
```diff
@@ -309,3 +313,84 @@ def post(self, user_id):
...
+            abort(400, message="There is no termination scheduled")
+
+        if not user.has_property('member'):
+            abort(400, message="User is not a member.")
+
+        cancel_task(task, user)
+
+        session.session.commit()
+
+        return "Membership termination cancelled."
```
[the remark this answers is not in the document]
Comment (ibot3) [reply web/api/v0/__init__.py]: There should be a rework of this behavior for the whole API in an other PR.

**Pull request merged**